pkg/rpcclient/actor/waiter.go

```diff
@@ -0,0 +1,201 @@
+		Unsubscribe(id string) error
+	}
+)
+
+// Wait allows to wait until transaction will be accepted to the chain. It can be
+// used as a wrapper for Send or SignAndSend and accepts transaction hash,
+// ValidUntilBlock value and an error. It returns transaction execution result
+// or an error if transaction wasn't accepted to the chain.
+func (a *Actor) Wait(h util.Uint256, vub uint32, err error) (*state.AppExecResult, error) {
+	if err != nil {
+		return nil, err
+	}
+	if wsW, ok := a.client.(WSWaiter); ok {
+		return a.waitWithWSWaiter(wsW, h, vub)
+	}
+	return a.waitWithSimpleWaiter(a.client, h, vub)
+}
+
+// waitWithSimpleWaiter waits until transaction is accepted to the chain and
+// returns its execution result or an error if it's missing from chain after
+// VUB block.
+func (a *Actor) waitWithSimpleWaiter(c SimpleWaiter, h util.Uint256, vub uint32) (*state.AppExecResult, error) {
+	const failedAttemptThreshold = 3
+	var (
+		currentHeight uint32
+		failedAttempt int
+		pollTime      = a.GetMillisecondsPerBlock() / 2
+	)
+	if pollTime == 0 {
+		pollTime = time.Second / 2
+	}
+	timer := time.NewTicker(pollTime)
+	defer timer.Stop()
+	for {
+		select {
+		case <-timer.C:
+			blockCount, err := c.GetBlockCount()
+			if err != nil {
+				failedAttempt++
+				if failedAttempt > failedAttemptThreshold {
+					return nil, fmt.Errorf("failed to retrieve block count: %w", err)
+				}
+				continue
+			}
+			failedAttempt = 0
+			if blockCount-1 > currentHeight {
+				currentHeight = blockCount - 1
+			}
+			t := trigger.Application
+			res, err := c.GetApplicationLog(h, &t)
+			if err == nil {
+				return &state.AppExecResult{
+					Container: h,
+					Execution: res.Executions[0],
+				}, nil
+			}
+			if currentHeight >= vub {
+				return nil, ErrTxNotAccepted
+			}
+
+		case <-c.Context().Done():
+			return nil, ErrContextDone
+		}
+	}
+}
+
+// waitWithWSWaiter waits until transaction is accepted to the chain and returns
+// its execution result or an error if it's missing from chain after VUB block.
+// It uses optimized web-socket waiter if possible.
+func (a *Actor) waitWithWSWaiter(c WSWaiter, h util.Uint256, vub uint32) (res *state.AppExecResult, waitErr error) {
+	var wsWaitErr error
+	defer func() {
+		if wsWaitErr != nil {
+			res, waitErr = a.waitWithSimpleWaiter(c, h, vub)
+			if waitErr != nil {
+				waitErr = fmt.Errorf("WS waiter error: %w, simple waiter error: %v", wsWaitErr, waitErr)
+			}
+		}
+	}()
+	rcvr := make(chan rpcclient.Notification)
+	defer func() {
+	drainLoop:
+		// Drain rcvr to avoid other notification receivers blocking.
+		for {
+			select {
+			case <-rcvr:
+			default:
+				break drainLoop
+			}
+		}
+		close(rcvr)
+	}()
+	blocksID, err := c.SubscribeForNewBlocksWithChan(nil, rcvr)
+	if err != nil {
+		wsWaitErr = fmt.Errorf("failed to subscribe for new blocks: %w", err)
+		return
+	}
+	defer func() {
+		err = c.Unsubscribe(blocksID)
+		if err != nil {
+			// TODO: ?
```

Comment:
@roman-khimov, we need to deside how handle this error. IMO, it's not that critical to ruine the whole execution result and return it from the `waitWithWSWaiter`. At the same time, failing unsbscription means that we're still subscribe to this topic and will receive notifications from the sever.

Reply:
> failing unsbscription means that we're still subscribe to this topic

Practically I'd say we're disconnected. Or we have a buggy server (in which case nothing really is reliable). Returning both AER and an error seems to be appropriate.

Reply:
Fixed.